controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go

```diff
@@ -300,55 +332,69 @@ func (r *KubeadmControlPlaneReconciler) initializeControlPlane(ctx context.Conte
 		ownerRef,
 	)
 	if err != nil {
-		return errors.Wrap(err, "Failed to generate bootstrap config")
+		errs = append(errs, errors.Wrap(err, "failed to generate bootstrap config"))
 	}
 
-	// Create the Machine
-	err = r.MachineGenerator.GenerateMachine(
-		ctx,
-		r.Client,
-		kcp.Namespace,
-		kcp.Name,
-		cluster.Name,
-		kcp.Spec.Version,
-		infraRef,
-		bootstrapRef,
-		generateKubeadmControlPlaneLabels(cluster.Name),
-		ownerRef,
-	)
-	if err != nil {
-		logger.Error(err, "Unable to create Machine")
-		r.recorder.Eventf(kcp, corev1.EventTypeWarning, "FailedCreateMachine", "Failed to create machine: %v", err)
+	// Only proceed to generating the Machine if we haven't encountered an error
+	if len(errs) == 0 {
+		err := r.MachineGenerator.GenerateMachine(
+			ctx,
+			r.Client,
+			kcp.Namespace,
+			kcp.Name,
+			cluster.Name,
+			kcp.Spec.Version,
+			infraRef,
+			bootstrapRef,
+			generateKubeadmControlPlaneLabels(cluster.Name),
+			ownerRef,
+		)
+		if err != nil {
+			errs = append(errs, errors.Wrap(err, "failed to create Machine"))
+		}
+	}
 
-		errs := []error{err}
+	// If we encountered any errors, attempt to clean up any dangling resources
+	if len(errs) > 0 {
+		if err := r.cleanupFromGeneration(ctx, infraRef, bootstrapRef); err != nil {
+			errs = append(errs, errors.Wrap(err, "failed to cleanup generated resources"))
+		}
+
+		return utilerrors.NewAggregate(errs)
+	}
+
+	return nil
+}
+
+func (r *KubeadmControlPlaneReconciler) cleanupFromGeneration(ctx context.Context, infraRef, bootstrapRef *corev1.ObjectReference) error {
```

Comment:
Can this take in `...*corev1.ObjectReference` and use the unstructured path for all of them in a loop, or is that being overly optimize-y?

Reply:
It seems fine, it looses a bit of granularity in the error messages, but I think it'll be fine since the wrapped error should give context there.